### DISCUSSION TITLE
Power Planning Script Confusion

### DISCUSSION THREAD

**Discussion by garfield118** [The-OpenROAD-Project/OpenROAD-flow-scripts#869] [General] [answered]
![image](https://user-images.githubusercontent.com/122075263/225269290-9c2a3a36-0af7-4f24-8728-36c5845a74c5.png)

![image](https://user-images.githubusercontent.com/122075263/225271391-b07eb6e2-2dca-41a0-bc55-77d986ec5167.png)

I have attached a snapshot of a macro at  floorplan stage of riscV32i.  I can see that the macros have Metal 4 power stripes on them but the power planning of the given openroad script do not seem to have any power stripes connecting them. Is there anyway I can check if the macros have not gotten any power stripes connecting them?

**Comment by vijayank88** [comment]
@gadfort fyi

**Comment by vijayank88** [comment]
Refer https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/pdn#define-a-macro-power-grid and `pdn.tcl` updated for the same.
@garfield118 did you checked final layout file?

**Reply by vijayank88** [reply]
@maliberty 
Can you please confirm, the pdn.tcl need to be updated?

**Reply by garfield118** [reply]
> Refer https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/src/pdn#define-a-macro-power-grid and `pdn.tcl` updated for the same. @garfield118 did you checked final layout file?

![image](https://user-images.githubusercontent.com/122075263/225539889-49fa0785-137d-4c6c-950d-d6463596d279.png)

This is the postroute odb. There are signal,clk and ground connection to them. But still the macros power connections stripes seems still missing to me(All the metal 4 stripes on the macro are not connected to any power stripes).

**Reply by vijayank88** [reply]
@garfield118 
How you resolved the issue?

**Reply by garfield118** [reply, marked as answer]
> @garfield118 How you resolved the issue?


 define_pdn_grid -name {core_macro} -voltage_domains {CORE} -macro -orient {R0 R180 MX MY} -halo {2.0 2.0 2.0 2.0} -default grid_over_boundary
 add_pdn_connect -grid {core_macro} -layers {M4 M5}


Yes, add this script to the /openRoad/pdn/grid_strategy-M2-M5-M7.tcl file. And the via's will be generated.

**Comment by maliberty** [comment]
I would need a test case to see what is wrong.  It isn't obvious from just the description here.

**Reply by vijayank88** [reply]
@maliberty 
For `nangate45` and `sky130` in [pdn.tcl](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/d0de724bf69e4d1b46f656be3c1920c5b52aed03/flow/platforms/nangate45/grid_strategy-M1-M4-M7.tcl#L24) `macro_grid` available to create power grid for macros.

In `ASAP7` only standard cell grid available and not defined any `macro_grid`. check this [grid_strategy](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/platforms/asap7/openRoad/pdn/grid_strategy-M2-M5-M7.tcl).

Is that need to be fixed or not?

**Reply by maliberty** [reply]
We didn't have any memories for asap7 so there was no need.  Feel free to create one if needed.

**Reply by vijayank88** [reply]
I am just referring this nangate45 macro grid.
https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/platforms/nangate45/grid_strategy-M1-M4-M7.tcl

For asap7
How to choose width, pitch and offset?
I have used M4 M5 for macro grid connection.

**Reply by vijayank88** [reply]
I made similar macro grid for ASAP7.
```
####################################
# grid for: CORE_macro_grid_1
####################################
define_pdn_grid -name {CORE_macro_grid_1} -voltage_domains {CORE} -macro -orient {R0 R180 MX MY} -halo {2.0 2.0 2.0 2.0} -cells {.*}
add_pdn_connect -grid {CORE_macro_grid_1} -layers {M4 M5}
####################################
# grid for: CORE_macro_grid_2
####################################
define_pdn_grid -name {CORE_macro_grid_2} -voltage_domains {CORE} -macro -orient {R90 R270 MXR90 MYR90} -halo {2.0 2.0 2.0 2.0} -cells {.*}
add_pdn_connect -grid {CORE_macro_grid_2} -layers {M4 M6}
```
But I don't find any connection with above update.

`asap7/riscv32i` has memories now.
